### PR TITLE
Add healthcheck for dockerised LTR model

### PR DESCRIPTION
**Pull request by bilbof**
This adds a connectivity and model status healthcheck for the model served by the tensorflow serving docker container on search machines.

Still to add: AWS Sagemaker status check

https://trello.com/c/Jgbd0TrY/1255-add-model-status-to-search-api-healthcheck

### CODE REVIEW

File: lib/learn_to_rank/ranker_status.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      response = HTTParty.get(url, options)
+      validate_response!(response)
+      model = JSON.parse(response.body).fetch("model_version_status", []).last
+      validate_model_healthy!(model)
+      true
+    rescue HTTParty::Error, SocketError, Timeout::Error => e
+      raise RankerServerError.new(message: e.message)
+    end
+
+    def tensorflow_serving_ip
```
Comment (barrucadu) [inline lib/learn_to_rank/ranker_status.rb]: Can we re-use the method from `LearnToRank::Ranker` here?

File: lib/learn_to_rank/ranker_status.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    def healthy?
+      @healthy && errors.none?
+    end
+
+  private
+
+    attr_reader :timeout
+
+    GOOD_STATES = %w(AVAILABLE InService Updating Creating SystemUpdating RollingBack).freeze
```
Comment (barrucadu) [inline lib/learn_to_rank/ranker_status.rb]: We might want to think more about the sagemaker states - eg, `RollingBack` could be cause for a warning, as it means something's gone wrong with the current model.

File: lib/learn_to_rank/ranker_status.rb (hunk 1 above)
Comment (bilbof) [reply lib/learn_to_rank/ranker_status.rb]: Yep, I've moved this into a helper.

File: lib/learn_to_rank/ranker_status.rb (hunk 2 above)
Comment (bilbof) [reply lib/learn_to_rank/ranker_status.rb]: OK. I've removed those AWS states for now, I'll re-add some later when I check Sagemaker health.

File: lib/learn_to_rank/ranker.rb [hunk 3]
```diff
@@ -89,14 +91,8 @@ def fetch_new_scores_from_serving(examples)
...
 
-    def tensorflow_serving_ip
-      if ENV["TENSORFLOW_SERVING_IP"].present?
-        ENV["TENSORFLOW_SERVING_IP"]
-      elsif %w(development).include? ENV["RACK_ENV"]
-        "reranker"
-      else
-        "0.0.0.0"
-      end
+    def log_error(error)
```
Comment (barrucadu) [inline lib/learn_to_rank/ranker.rb]: I think you've gained an unused function through a rebase

Comment (bilbof) [reply lib/learn_to_rank/ranker.rb]: Thanks for spotting that - I've removed it!

File: spec/integration/app/healthcheck_spec.rb [hunk 4]
```diff
@@ -27,6 +31,28 @@
...
   end
 
+  describe "#reranker_healthcheck check" do
+    # We only check for cannot connect because govuk_app_config has tests for this
+    context "when reranker healthcheck fails" do
+      before do
+        stub_ranker_container_doesnt_exist
+      end
+
+      it "returns a critical status" do
```
Comment (barrucadu) [inline spec/integration/app/healthcheck_spec.rb]: "returns a warning status" - or should the healthcheck actually be critical?

I'm leaning towards a warning being the right status, as search isn't *broken* when the reranker is down, users just get a degraded experience.

**Review by barrucadu — APPROVED** [review]
Looks good, there's just one misnamed test

File: spec/integration/app/healthcheck_spec.rb (hunk 4 above)
Comment (bilbof) [reply spec/integration/app/healthcheck_spec.rb]: Thanks, I've fixed that.

**Pull request merged**